storage: check for uninitialized raft group

irfansharif opened the pull request:
Fixes #16184.
Fixes #16185.
Fixes #16193.

Check to see if replica hasn't been destroyed before updating proposal
quota, the resulting nil RaftStatus causes panics if a leader replica is
being destroyed.

irfansharif committed:
storage: check for uninitialized raft group

Fixes #16184.
Fixes #16185.
Fixes #16193.

Check to see if replica hasn't been destroyed before updating proposal
quota.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16187)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -905,6 +905,10 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 	// ourselves via looking at MsgAppResp messages would be overkill. Perhaps
 	// another accessor on RawNode.
 	status := r.raftStatusRLocked()
+	if status == nil {
```
should this be higher up? Or, in case we don't have a Raft group, what happens in the code above this?

irfansharif replied:
it should be fine here, if we get to this codepath if you look we're not actually doing anything above. The mistake here was assuming proper initialization but as is in the case of the failures above, does not happen if the replica is destroyed ([here](https://github.com/irfansharif/cockroach/blob/79d6e2207fcb7c2977591ae856a6b0fbcda6a0d3/pkg/storage/replica.go#L2990-L2992), then [here](https://github.com/irfansharif/cockroach/blob/79d6e2207fcb7c2977591ae856a6b0fbcda6a0d3/pkg/storage/replica.go#L457) never reaching [here](https://github.com/irfansharif/cockroach/blob/79d6e2207fcb7c2977591ae856a6b0fbcda6a0d3/pkg/storage/replica.go#L486)).

tschottdorf replied:
If the replica got destroyed, wouldn't the correct next step be destroying the quota pool as well? Otherwise (ok, and this is probably hard to get in practice) a replica could be destroyed but some goroutines are still waiting for quota and will never get it?

tamird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-Kl9pDcDm2sS8OcJR7bW-r1-908:-KlAHlJFsM4d3aYU70yE:b8ez4oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/79d6e2207fcb7c2977591ae856a6b0fbcda6a0d3/pkg/storage/replica.go#L908)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

If the replica got destroyed, wouldn't the correct next step be destroying the quota pool as well? Otherwise (ok, and this is probably hard to get in practice) a replica could be destroyed but some goroutines are still waiting for quota and will never get it?
</blockquote></details>

This certainly needs a comment. For one, I can't tell what this means, particularly since the above code block prescribes that this replica is a leader or follower, both of which contradict this clause (which implies that the replica is being GC'ed).

Naively, I echo @tschottdorf's suggestion that this should result in some resource cleanup, though perhaps that's already being done elsewhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-Kl9pDcDm2sS8OcJR7bW-r1-908:-KlAdo_X80no-Yg5CPX3:b-7estr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/79d6e2207fcb7c2977591ae856a6b0fbcda6a0d3/pkg/storage/replica.go#L908)):*
> replica could be destroyed but some goroutines are still waiting for quota and will never get it?

This is plausible, yes. Added.

>  I can't tell what this means, particularly since the above code block prescribes that this replica is a leader or follower, both of which contradict this clause (which implies that the replica is being GC'ed).

I was actually incorrect in saying this was due to improper initialization, the failure here occurs when a leader replica is being removed/destroyed (as in `Store.removeReplicaImpl`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 899 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-KlBOjEnZerd8ZrPhums:-KlBOjEoKsEQsbaHVx5i:b-eu0chb) ([raw file](https://github.com/cockroachdb/cockroach/blob/03302c551a4deb9353d8667ee4d6dd39b98cda0f/pkg/storage/replica.go#L899)):*
> ```Go
> 		// We're a follower.
> 		return
> 	}
> ```

how does the new case relate to this if statement? is it equivalent as an `else if` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 899 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-KlBOjEnZerd8ZrPhums:-KlBm1d69v76TXa6jAtH:bg92lcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/03302c551a4deb9353d8667ee4d6dd39b98cda0f/pkg/storage/replica.go#L899)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how does the new case relate to this if statement? is it equivalent as an `else if` here?
</blockquote></details>

you _can_ make this into an `else if` but the if statement here deals with leadership transitions or the lack thereof as seen by followers, old leaders and new. The fall through case when the replication is _still_ the leader is where we need to close the quota pool if destroyed, so it should be fine as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
okay to merge?

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -900,11 +900,25 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 	// another accessor on RawNode.
 	status := r.raftStatusRLocked()
+	if status == nil {
+		log.Fatal(ctx, "leader with nil internalRaftGroup")
```
While this comment is true, what you're really complaining about is "leader with nil RaftStatus".

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -900,11 +900,25 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 
+	// We need to check if the replica is being destroyed and if so, close the
+	// quota pool and unblock ongoing quota acquisition goroutines (if any).
+	if r.mu.destroyed != nil {
```
Strikes me as odd that we need to do this here, and there is still a potential problem here that leaves clients hanging if the replica gets destroyed but this code path is never taken. Having taken a rough look at the ways in which replicas can get destroyed unfortunately convinced me that you shouldn't be burdened with addressing that as well - it's a bit of a mess.

tschottdorf commented:
LGTM

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 905 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-Kl9pDcDm2sS8OcJR7bW-r2-905:-KlM7eUZQOlhDPNMGRIo:bg12vwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/03302c551a4deb9353d8667ee4d6dd39b98cda0f/pkg/storage/replica.go#L905)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Strikes me as odd that we need to do this here, and there is still a potential problem here that leaves clients hanging if the replica gets destroyed but this code path is never taken. Having taken a rough look at the ways in which replicas can get destroyed unfortunately convinced me that you shouldn't be burdened with addressing that as well - it's a bit of a mess.
</blockquote></details>

Yeah, it seems weird that we could "still be the leader" on a replica that's destroyed. Should we be closing proposalQuota in `Store.removeReplicaImpl`? I think there are only three ways a replica can be destroyed (the normal case in `Store.removeReplicaImpl`, `maybeSetCorrupt`, and merges), so it shouldn't be too bad to ensure that we clear our leader status and close the proposalQuota whenever the replica is destroyed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 905 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-Kl9pDcDm2sS8OcJR7bW-r2-905:-KlM9zkml7QUBWeZ32Ym:b-vjhw0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/03302c551a4deb9353d8667ee4d6dd39b98cda0f/pkg/storage/replica.go#L905)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, it seems weird that we could "still be the leader" on a replica that's destroyed. Should we be closing proposalQuota in `Store.removeReplicaImpl`? I think there are only three ways a replica can be destroyed (the normal case in `Store.removeReplicaImpl`, `maybeSetCorrupt`, and merges), so it shouldn't be too bad to ensure that we clear our leader status and close the proposalQuota whenever the replica is destroyed. 
</blockquote></details>

this is a half measure. short of clearing the leader status before destroying the replica (or closing the pool before/after we destroy a replica) this patch should address the flakiness people are sure to run into tomorrow. added a TODO to follow up on points mentioned here.

---

*[pkg/storage/replica.go, line 919 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187#-Kl9pDcDm2sS8OcJR7bW-r2-919:-KlMA2kNdAqrDPegwt7b:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/03302c551a4deb9353d8667ee4d6dd39b98cda0f/pkg/storage/replica.go#L919)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

While this comment is true, what you're really complaining about is "leader with nil RaftStatus".
</blockquote></details>

Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16187)*
<!-- Sent from Reviewable.io -->
